// ng-sidebar/lib/sidebar-container.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var sidebar_component_1 = require("./sidebar.component");
var utils_1 = require("./utils");
// Based on https://github.com/angular/material2/tree/master/src/lib/sidenav
var SidebarContainer = (function () {
    function SidebarContainer(_ref, _el) {
        this._ref = _ref;
        this._el = _el;
        this.allowSidebarBackdropControl = true;
        this.animate = true;
        this.showBackdrop = false;
        this.showBackdropChange = new core_1.EventEmitter();
    }
    SidebarContainer.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (!utils_1.isBrowser()) {
            return;
        }
        this._onToggle();
        this._subscribe();
        this._sidebars.changes.subscribe(function () {
            _this._unsubscribe();
            _this._subscribe();
        });
    };
    SidebarContainer.prototype.ngOnChanges = function (changes) {
        if (!utils_1.isBrowser()) {
            return;
        }
        if (changes['showBackdrop']) {
            this.showBackdropChange.emit(changes['showBackdrop'].currentValue);
        }
    };
    SidebarContainer.prototype.ngOnDestroy = function () {
        if (!utils_1.isBrowser()) {
            return;
        }
        this._unsubscribe();
    };
    /**
     * @internal
     *
     * Computes `margin` value to push page contents to accommodate open sidebars as needed.
     *
     * @return {CSSStyleDeclaration} margin styles for the page content.
     */
    SidebarContainer.prototype._getContentStyles = function () {
        var _this = this;
        var left = 0, right = 0, top = 0, bottom = 0;
        if (this._sidebars) {
            this._sidebars.forEach(function (sidebar) {
                if (!sidebar) {
                    return;
                }
                if (sidebar.mode === 'slide') {
                    var transformStyle = null;
                    if (sidebar.opened) {
                        var isLeftOrTop = sidebar.position === 'left' || sidebar.position === 'top';
                        var isLeftOrRight = sidebar.position === 'left' || sidebar.position === 'right';
                        var transformDir = isLeftOrRight ? 'X' : 'Y';
                        var transformAmt = "" + (isLeftOrTop ? '' : '-') + (isLeftOrRight ? sidebar._width : sidebar._height);
                        transformStyle = "translate" + transformDir + "(" + transformAmt + "px)";
                    }
                    _this._el.nativeElement.style.transform = transformStyle;
                }
                if ((sidebar.mode === 'push' && sidebar.opened) || sidebar.mode === 'dock') {
                    switch (sidebar.position) {
                        case 'left':
                            left = Math.max(left, sidebar._width);
                            break;
                        case 'right':
                            right = Math.max(right, sidebar._width);
                            break;
                        case 'top':
                            top = Math.max(top, sidebar._height);
                            break;
                        case 'bottom':
                            bottom = Math.max(bottom, sidebar._height);
                            break;
                    }
                }
            });
        }
        return {
            margin: top + "px " + right + "px " + bottom + "px " + left + "px"
        };
    };
    /**
     * @internal
     *
     * Closes sidebars when the backdrop is clicked, if they have the
     * `closeOnClickBackdrop` option set.
     */
    SidebarContainer.prototype._onBackdropClicked = function () {
        this._sidebars.forEach(function (sidebar) {
            if (sidebar.opened && sidebar.showBackdrop && sidebar.closeOnClickBackdrop) {
                sidebar.close();
            }
        });
    };
    /**
     * @internal
     *
     * Subscribes from all sidebar events to react properly.
     */
    SidebarContainer.prototype._subscribe = function () {
        var _this = this;
        if (this._sidebars) {
            this._sidebars.forEach(function (sidebar) {
                if (!sidebar) {
                    return;
                }
                sidebar.onOpenStart.subscribe(function () { return _this._onToggle(); });
                sidebar.onOpened.subscribe(function () { return _this._markForCheck(); });
                sidebar.onCloseStart.subscribe(function () { return _this._onToggle(); });
                sidebar.onClosed.subscribe(function () { return _this._markForCheck(); });
                sidebar.onModeChange.subscribe(function () { return _this._markForCheck(); });
                sidebar.onPositionChange.subscribe(function () { return _this._markForCheck(); });
                sidebar._onRerender.subscribe(function () { return _this._markForCheck(); });
            });
        }
    };
    /**
     * @internal
     *
     * Unsubscribes from all sidebars.
     */
    SidebarContainer.prototype._unsubscribe = function () {
        if (this._sidebars) {
            this._sidebars.forEach(function (sidebar) {
                if (!sidebar) {
                    return;
                }
                sidebar.onOpenStart.unsubscribe();
                sidebar.onOpened.unsubscribe();
                sidebar.onCloseStart.unsubscribe();
                sidebar.onClosed.unsubscribe();
                sidebar.onModeChange.unsubscribe();
                sidebar.onPositionChange.unsubscribe();
                sidebar._onRerender.unsubscribe();
            });
        }
    };
    /**
     * @internal
     *
     * Triggers change detection to recompute styles.
     */
    SidebarContainer.prototype._markForCheck = function () {
        this._ref.markForCheck();
    };
    /**
     * @internal
     *
     * Check if we should show the backdrop when a sidebar is toggled.
     */
    SidebarContainer.prototype._onToggle = function () {
        if (this._sidebars && this.allowSidebarBackdropControl) {
            var hasOpen = false;
            var _sidebars = this._sidebars.toArray();
            for (var i = 0; i < _sidebars.length; i++) {
                var sidebar = _sidebars[i];
                // Show backdrop if a single open sidebar has it set
                if (sidebar.opened && sidebar.showBackdrop) {
                    hasOpen = true;
                    break;
                }
            }
            this.showBackdrop = hasOpen;
            this.showBackdropChange.emit(hasOpen);
        }
        this._markForCheck();
    };
    return SidebarContainer;
}());
SidebarContainer.decorators = [
    { type: core_1.Component, args: [{
                selector: 'ng-sidebar-container',
                template: "\n    <ng-content select=\"ng-sidebar\"></ng-content>\n\n    <div *ngIf=\"showBackdrop\"\n      aria-hidden=\"true\"\n      class=\"ng-sidebar__backdrop\"\n      [ngClass]=\"backdropClass\"\n      (click)=\"_onBackdropClicked()\"></div>\n\n    <div class=\"ng-sidebar__content\"\n      [ngClass]=\"sidebarContentClass\"\n      [ngStyle]=\"_getContentStyles()\">\n      <ng-content></ng-content>\n    </div>\n  ",
                styles: ["\n    ng-sidebar-container {\n      box-sizing: border-box;\n      display: block;\n      position: relative;\n    }\n\n      ng-sidebar-container.ng-sidebar-container--animate {\n        -webkit-transition: -webkit-transform 0.3s cubic-bezier(0, 0, 0.3, 1);\n        transition: transform 0.3s cubic-bezier(0, 0, 0.3, 1);\n      }\n\n    .ng-sidebar__backdrop {\n      background: #000;\n      height: 100%;\n      left: 0;\n      opacity: 0.75;\n      pointer-events: auto;\n      position: fixed;\n      top: 0;\n      width: 100%;\n      z-index: 99999998;\n    }\n\n    .ng-sidebar__content {\n      display: block;\n      height: 100%;\n      overflow: auto;\n    }\n\n      .ng-sidebar-container--animate .ng-sidebar__content {\n        -webkit-transition: margin 0.3s cubic-bezier(0, 0, 0.3, 1);\n        transition: margin 0.3s cubic-bezier(0, 0, 0.3, 1);\n      }\n  "],
                host: {
                    '[class.ng-sidebar-container--animate]': 'animate'
                },
                changeDetection: core_1.ChangeDetectionStrategy.OnPush,
                encapsulation: core_1.ViewEncapsulation.None
            },] },
];
/** @nocollapse */
SidebarContainer.ctorParameters = function () { return [
    { type: core_1.ChangeDetectorRef, },
    { type: core_1.ElementRef, },
]; };
SidebarContainer.propDecorators = {
    'sidebarContentClass': [{ type: core_1.Input },],
    'backdropClass': [{ type: core_1.Input },],
    'allowSidebarBackdropControl': [{ type: core_1.Input },],
    'animate': [{ type: core_1.Input },],
    'showBackdrop': [{ type: core_1.Input },],
    'showBackdropChange': [{ type: core_1.Output },],
    '_sidebars': [{ type: core_1.ContentChildren, args: [sidebar_component_1.Sidebar,] },],
};
exports.SidebarContainer = SidebarContainer;
